File: backend/src/Alife.Application/IdentityAccess/IdentityPathPolicy.cs
namespace Alife.Application.IdentityAccess;

public static class IdentityPathPolicy
{
    public static string NormalizeReturnPath(string? value)
    {
        var candidate = value?.Trim() ?? string.Empty;
        if (candidate.Length == 0 || candidate.Length > 1000 ||
            !candidate.StartsWith('/') || candidate.StartsWith("//", StringComparison.Ordinal) ||
            candidate.Contains('\\') || candidate.Any(char.IsControl) ||
            Uri.TryCreate(candidate, UriKind.Absolute, out _))
        {
            return string.Empty;
        }

        var path = candidate.Split('?', '#')[0];
        if (path.Equals("/onboarding", StringComparison.OrdinalIgnoreCase) ||
            path.StartsWith("/onboarding/", StringComparison.OrdinalIgnoreCase) ||
            path.StartsWith("/activate/", StringComparison.OrdinalIgnoreCase) ||
            path.StartsWith("/join/", StringComparison.OrdinalIgnoreCase) ||
            path.StartsWith("/application/", StringComparison.OrdinalIgnoreCase) ||
            path.Equals("/internal/alpha-login", StringComparison.OrdinalIgnoreCase))
        {
            return string.Empty;
        }

        return candidate;
    }
}

File: cloudflare/alife-app/src/services/identityPathPolicy.ts
const NON_RETURNABLE = [
  '/onboarding',
  '/internal/alpha-login',
]

export const normalizeIdentityReturnPath = (value: string | null | undefined) => {
  const candidate = value?.trim() ?? ''
  if (
    !candidate.startsWith('/') ||
    candidate.startsWith('//') ||
    candidate.includes('\\') ||
    /[\u0000-\u001f\u007f]/.test(candidate)
  ) {
    return ''
  }

  const pathname = candidate.split(/[?#]/, 1)[0].toLowerCase()
  if (
    NON_RETURNABLE.includes(pathname) ||
    pathname.startsWith('/onboarding/') ||
    pathname.startsWith('/activate/') ||
    pathname.startsWith('/join/') ||
    pathname.startsWith('/application/')
  ) {
    return ''
  }

  return candidate
}

export const buildOnboardingLocation = (returnPath: string) => {
  const safe = normalizeIdentityReturnPath(returnPath)
  return safe ? `/onboarding?returnTo=${encodeURIComponent(safe)}` : '/onboarding'
}
